File: androidDemo/src/main/java/com/novage/p2pml/demo/logging/P2PFileLogSink.kt
package com.novage.p2pml.demo.logging

import android.content.Context
import android.content.Intent
import android.os.Build
import androidx.core.content.FileProvider
import com.novage.p2pml.api.logging.LogLevel
import com.novage.p2pml.api.logging.P2PLogger
import com.novage.p2pml.api.logging.P2PLogging
import java.io.File
import java.io.PrintWriter
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.concurrent.Executors

/**
 * Persists library logs to a per-session file so they can be pulled off the device later
 * (share sheet, `adb pull`, or Android Studio Device Explorer). Chains in front of the
 * library's default sink, so Logcat keeps working.
 */
object P2PFileLogSink {
    private const val LOG_DIR = "p2pml-logs"
    private const val MAX_SESSION_FILES = 5

    private val writeExecutor = Executors.newSingleThreadExecutor()

    // Confined to writeExecutor's single thread (SimpleDateFormat is not thread-safe).
    private val lineTimeFormat = SimpleDateFormat("HH:mm:ss.SSS", Locale.US)
    private var writer: PrintWriter? = null

    private var logFile: File? = null

    fun install(context: Context) {
        if (logFile != null) return

        val file = createSessionFile(context) ?: return
        logFile = file

        // Queued before the sink is chained, so the single-threaded executor guarantees
        // the writer exists by the time the first writeEntry task runs.
        writeExecutor.execute {
            writer = PrintWriter(file.bufferedWriter())
            writer?.println(
                "--- P2PML demo session, ${Build.MANUFACTURER} ${Build.MODEL}, API ${Build.VERSION.SDK_INT} ---"
            )
        }

        val platformSink = P2PLogging.sink
        P2PLogging.sink = P2PLogger { level, tag, message, throwable ->
            platformSink?.log(level, tag, message, throwable)
            writeEntry(System.currentTimeMillis(), level, tag, message, throwable)
        }
    }

    /** Opens a share sheet for the current session's log file. */
    fun share(context: Context) {
        val file = logFile ?: return

        val uri = FileProvider.getUriForFile(context, "${context.packageName}.fileprovider", file)
        val intent = Intent(Intent.ACTION_SEND).apply {
            type = "text/plain"
            putExtra(Intent.EXTRA_STREAM, uri)
            addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
        }
        context.startActivity(Intent.createChooser(intent, "Share P2P logs"))
    }

    private fun createSessionFile(context: Context): File? = runCatching {
        val dir = File(context.getExternalFilesDir(null) ?: context.filesDir, LOG_DIR)
        dir.mkdirs()

        // Timestamped names sort lexicographically — prune all but the newest sessions.
        dir.listFiles()
            ?.sortedByDescending { it.name }
            ?.drop(MAX_SESSION_FILES - 1)
            ?.forEach { it.delete() }

        val sessionStamp = SimpleDateFormat("yyyy-MM-dd_HH-mm-ss", Locale.US).format(Date())
        File(dir, "p2pml-$sessionStamp.log")
    }.getOrNull()

    private fun writeEntry(
        timestampMs: Long,
        level: LogLevel,
        tag: String,
        message: String,
        throwable: Throwable?
    ) {
        writeExecutor.execute {
            writer?.apply {
                val time = lineTimeFormat.format(Date(timestampMs))
                println("$time ${level.name.first()} [${tag.removePrefix("P2PML~")}] $message")
                throwable?.let { println(it.stackTraceToString()) }
                flush()
            }
        }
    }
}

File: iosDemo/iosDemo/Logging/P2PFileLogSink.swift
import Foundation
import P2PML
import UIKit

/// Persists library logs to a per-session file in Documents/p2pml-logs so they can be
/// pulled off the device later (Files app, Finder device pane, or the share sheet).
/// Chains in front of the library's default sink, so the Xcode console keeps working.
enum P2PFileLogSink {
    private static let logDirName = "p2pml-logs"
    private static let maxSessionFiles = 5

    private static let writeQueue = DispatchQueue(label: "com.novage.p2pml.demo.logsink")
    private static var fileHandle: FileHandle?

    private(set) static var currentLogURL: URL?

    private static let lineTimeFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "HH:mm:ss.SSS"
        return formatter
    }()

    static func install() {
        guard currentLogURL == nil, let url = createSessionFile() else { return }
        currentLogURL = url

        let device = UIDevice.current
        let sessionHeader = "--- P2PML demo session, \(device.model), iOS \(device.systemVersion) ---"
        // Queued before the sink is chained, so the serial queue guarantees the
        // handle exists by the time the first writeEntry task runs.
        writeQueue.async {
            fileHandle = try? FileHandle(forWritingTo: url)
            appendLine(sessionHeader)
        }

        let platformSink = P2PLogging.shared.sink
        P2PLogging.shared.sink = ChainedSink(platformSink: platformSink)
    }

    fileprivate static func writeEntry(level: LogLevel, tag: String, message: String, throwable: KotlinThrowable?) {
        let timestamp = Date()
        writeQueue.async {
            let time = lineTimeFormatter.string(from: timestamp)
            let shortTag = tag.replacingOccurrences(of: "P2PML~", with: "")
            var line = "\(time) \(level.name.prefix(1)) [\(shortTag)] \(message)"
            if let throwable {
                line += "\n\(throwable.description)"
            }
            appendLine(line)
        }
    }

    private static func appendLine(_ line: String) {
        guard let data = (line + "\n").data(using: .utf8) else { return }
        // The throwing FileHandle APIs: the legacy ones raise uncatchable
        // NSExceptions if the file disappears (it is user-deletable via the Files app).
        _ = try? fileHandle?.seekToEnd()
        try? fileHandle?.write(contentsOf: data)
    }

    private static func createSessionFile() -> URL? {
        let fileManager = FileManager.default
        guard let documents = fileManager.urls(for: .documentDirectory, in: .userDomainMask).first else {
            return nil
        }

        let dir = documents.appendingPathComponent(logDirName, isDirectory: true)
        try? fileManager.createDirectory(at: dir, withIntermediateDirectories: true)

        // Timestamped names sort lexicographically — prune all but the newest sessions.
        if let existing = try? fileManager.contentsOfDirectory(at: dir, includingPropertiesForKeys: nil) {
            existing
                .sorted { $0.lastPathComponent > $1.lastPathComponent }
                .dropFirst(maxSessionFiles - 1)
                .forEach { try? fileManager.removeItem(at: $0) }
        }

        let sessionFormatter = DateFormatter()
        sessionFormatter.dateFormat = "yyyy-MM-dd_HH-mm-ss"
        let url = dir.appendingPathComponent("p2pml-\(sessionFormatter.string(from: Date())).log")
        fileManager.createFile(atPath: url.path, contents: nil)
        return url
    }
}

private final class ChainedSink: NSObject, P2PLogger {
    private let platformSink: P2PLogger?

    init(platformSink: P2PLogger?) {
        self.platformSink = platformSink
    }

    func log(level: LogLevel, tag: String, message: String, throwable: KotlinThrowable?) {
        platformSink?.log(level: level, tag: tag, message: message, throwable: throwable)
        P2PFileLogSink.writeEntry(level: level, tag: tag, message: message, throwable: throwable)
    }
}
